### ISSUE TITLE
CtrlPTag jumps to wrong name

### ISSUE THREAD

**Issue by osa1** [ctrlpvim/ctrlp.vim#92]
Here's an example, I have these entries in my tags file:

```
Benchmark	Criterion/Types.hs	331 ;" t
Benchmark	Criterion/Types.hs	333 ;" C
```

When I select one of them in `CtrlPTag` buffer, both of them jump to line 333, instead of first one jumping to 331.

Speculation: I think what's happening is that if I have multiple entries with same names, it jumps to only one of them, no matter which one I select in `CtrlPTag` buffer.

Any ideas?

**Comment by mattn** [comment]
Could you please try following patch?

```diff
diff --git a/autoload/ctrlp/tag.vim b/autoload/ctrlp/tag.vim
index fa8c732..fbdc7c9 100644
--- a/autoload/ctrlp/tag.vim
+++ b/autoload/ctrlp/tag.vim
@@ -100,7 +100,7 @@ fu! ctrlp#tag#accept(mode, str)
 		\ 'v': ['vs', 'vert stj'],
 		\ 'e': ['', 'tj'],
 		\ }
-	let utg = fdcnt[3] < 2 && fdcnt[0] == 1 && fdcnt[1] == 1
+	let utg = fdcnt[3] < 2 && fdcnt[0] == 1
 	let cmd = !fdcnt[0] || utg ? cmds[a:mode][0] : cmds[a:mode][1]
 	let cmd = a:mode == 'e' && ctrlp#modfilecond(!&aw)
 		\ ? ( cmd == 'tj' ? 'stj' : 'sp' ) : cmd
@@ -111,7 +111,7 @@ fu! ctrlp#tag#accept(mode, str)
 		en
 		let save_cst = &cst
 		set cst&
-		cal feedkeys(":".( utg ? fdcnt[2] : "" )."ta ".tg."\r", 'nt')
+		cal feedkeys(":".( utg ? fdcnt[3] : "" )."ta ".tg."\r", 'nt')
 		let &cst = save_cst
 	el
 		let ext = ""
```

**Comment by osa1** [comment]
No, it's still broken. Interestingly, now what happens is that it sometimes jumps to line 331, sometimes to 333. It's hard to describe but I think if you create a tag file with multiple entries with same symbols and play around with it you'll quickly see how it's broken.

**Comment by mattn** [comment]
After applying this patch, I couldn't reproduce this issue. what are differences

**Comment by osa1** [comment]
It's hard to figure out when exactly this is not working. Here's a way to reproduce the wrong behavior:

* Run vim, jump to to `Benchmark` tag on line 333. It jumps to correct place.
* Jump to `Benchmark` tag on line 331 while on line 333, it doesn't jump to 331, stays on 333.
* Move cursor to some place else in the same file, try again. Still doesn't work. Always jumps to 333.
* Switch to another buffer/file, try jumping to 331, this time it works.
* Now try to jump to 333, again, it works.
* Now try to jump to 331, doesn't work, stays on 333.
* Switch to another buffer again, this time try to jump to 333. It jumps to 331.

This is clearly broken.

The code base I'm using: https://github.com/bos/criterion
Tags file: https://gist.github.com/osa1/4475c438a6d42c5dffb4

**Comment by ralfthewise** [comment]
I'm also having this issue, will try the patch later tonight

**Comment by mattn** [comment]
@osa1 could you please try this again?

**Comment by osa1** [comment]
It's still broken, but in a different way. In the example I gave above, previously it was sometimes jumping to line 331 and sometimes jumping to line 333, now it's only jumping to line 333.

@mattn could you try to reproduce my case? That should help debugging. I gave the tags file and git repository already(see my previous comment).

**Comment by mattn** [comment]
Yes, i did. But couldn't reproduce.

**Comment by osa1** [comment]
This is ridiculous. @mattn please revert that patch, it ruined everything. Things are a lot worse when you have different tags with same case-insensitive names. Previously it was working at least when tags are in different files.

Steps the reproduce new, even worse behavior:

- Clone the repo above.
- Download the tags file I linked above.
- You should have 3 `benchmark` tags, two of them are in same file and capitalized the same way(i.e. `Benchmark`), other one is `benchmark` and in different file.
- All of them jumps to same tag.

If you can't reproduce that, try a few times. It seems like something is triggering this behavior.

**Comment by osa1** [comment]
I'm wondering how are you testing this.

**Comment by osa1** [comment]
Correction: This isn't a new behavior, it looks like it was already broken in `5ab5b0d`. I thought it was working before when tags are in different files but I think I'm wrong about that.

**Comment by osa1** [comment]
Here's a demo for you:
![gifrecord_2015-08-08_121024](https://cloud.githubusercontent.com/assets/448274/9151152/c637ee50-3dc6-11e5-8a29-93c01fb28402.gif)

**Comment by mattn** [comment]
> please revert that patch
Which patch do you mean?

**Comment by osa1** [comment]
@mattn never mind that. What I meant is it's still broken.

Did the demo help? Can you reproduce it now?

**Comment by mattn** [comment]
I'm looking code. And I'm confusing why the part of jump-tag is like that.

**Comment by osa1** [comment]
I recently tried debugging this. The problem I think is that `:tj <name>` asks for a number, if there are multiple tags with same name. But these numbers for different tags with same names aren't fixed, for example: when I do `:tj benchmark`, 1 is for `benchmark`, 2 is for `Benchmark` and 3 is for the last `Benchmark`. But when I do `:tj Benchmark` this list is in different order, and I don't think `s:findcount()` in `tag.vim` is taking that into account. I'll keep debugging this sometime later.

**Comment by ludovicchabant** [comment]
I was able to reproduce the bug with an old copy of CtrlP, and the bug is gone for my test case after @mattn's patch.

FYI, I didn't use the `criterion` codebase to reproduce because it's written in Haskell and `ctags` doesn't support that language out of the box, which means installing a different tag generator, which adds more moving parts to this thing.... so I just wrote a little fake C program that has multiple functions with the same name (some with exact same name, some with upper/lowercase differences). I tried navigating to those from either the same file, and from another file.

**Comment by Cimbali** [comment]
(sorry for the late reply but I'm looking into CtrlPTag a little)

@osa1 It appears that your tags file was ill-formed according to vim, thus when it gave the `taglist()` output, it did not parse commands and extra fields correctly, instead bundling it together as the command.

If you replace all occurences of ` ;" ` with `;"\t` in that tags file, it works fine. Which explains why other people can't reproduce with other languages. See also [the format specification](http://ctags.sourceforge.net/FORMAT).
